src/net/network_manager.cpp

```diff
@@ -60,4 +83,63 @@ void NetworkManager::run_service() {
             std::cerr << "run_service: io_service error\n";
         }
     }
-}
\ No newline at end of file
+}
+
+void NetworkManager::update() {
+    if (server)
+        update_server();
+    if (client)
+        update_client();
+}
+
+void NetworkManager::update_server() {
+    for (auto& client : server->read_all_messages()) {
+        for (auto& msg : client.second) {
+            switch (msg.message_type_case()) {
+            case proto::ClientMessage::MessageTypeCase::kBuildRequest: {
+                proto::Construct construct = msg.build_request();
+                events::build::ConstructData cd;
+                to_construct_data(construct, cd);
+                events::build::try_build_event(cd);
+                break;
+            }
+            default:
+                break;
+            }
+        }
+    }
+}
+
+void NetworkManager::update_client() {
+    proto::ServerMessage msg;
+    while (client->read_message(&msg)) {
+        switch (msg.message_type_case()) {
+        case proto::ServerMessage::MessageTypeCase::kBuildUpdate: {
+            if (!msg.status()) {
+                // TODO: construct placement failed. notify player?
+                break;
+            }
+
+            proto::Construct construct = msg.build_update();
+            events::build::ConstructData cd;
+            to_construct_data(construct, cd);
+            events::build::update_build_event(cd);
+        }
+        default:
+            break;
+        }
+    }
+}
+
+void NetworkManager::to_construct(events::build::ConstructData& cd, proto::Construct* c) {
+    c->set_type(cd.type);
+    c->set_x(cd.x);
+    c->set_z(cd.z);
+}
+
+void NetworkManager::to_construct_data(proto::Construct& c, events::build::ConstructData& cd) {
+    cd.type = static_cast<ConstructType>(c.type());
+    cd.x = c.x();
+    cd.z = c.z();
+}
```

Comment:
This is because the events use ConstructData (custom struct) while protobuf uses proto::Construct.  The alternative is to only use proto::Construct everywhere instead, which sounds alright.